client/core/device.go

```diff
@@ -48,6 +50,63 @@ type TLSConfig struct {
 	GetCertificateAuthorities GetCertificateAuthoritiesFunc
 }
 
+type conn struct {
+	mutex *semaphore.Weighted
+	c     atomic.Value // *coap.ClientCloseHandler
```

Comment:
it's a bit unusual to have both an `atomic` variable and a synchronization primitive

Reply:
Mainly it is used for synchronize dial and remove  call. Only one dial can be called to one endpoint. And if dial is progress then remove function do nothing.